### PR TITLE
Added support for image prefix in setup & upgrade command

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

File: space-cli/cmd/modules/operations/commands.go [hunk 1]
```diff
@@ -145,6 +147,7 @@ func Commands() []*cobra.Command {
 	if err != nil {
 		_ = utils.LogError("Unable to bind lag ('cluster-name') to environment variables", nil)
 	}
+	setup.Flags().StringP("image-prefix", "", "", "Prefix to use for providing custom image names")
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/commands.go:150]: Set the default value to `spaceuptech`

File: space-cli/cmd/modules/operations/commands.go [hunk 2]
```diff
@@ -160,11 +163,15 @@ func Commands() []*cobra.Command {
...
 			}
+			if err := viper.BindPFlag("image-prefix", cmd.Flags().Lookup("image-prefix")); err != nil {
+				_ = utils.LogError("Unable to bind the flag ('image-prefix')", nil)
+			}
 		},
 		RunE: actionUpgrade,
 	}
 	upgrade.Flags().StringP("cluster-name", "", "default", "The name of space-cloud cluster")
 	upgrade.Flags().StringP("version", "", "default", "version to use for upgrade")
+	upgrade.Flags().StringP("image-prefix", "", "", "Prefix to use for providing custom image names")
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/commands.go:174]: Set the default value to `spaceuptech`

File: space-cli/cmd/utils/operations.go [hunk 3]
```diff
@@ -93,6 +95,22 @@ func ChangeSelectedAccount(clusterName string) error {
 	return nil
 }
 
+//GetSCImageName get the sc image name and add the image prefix when required
+func GetSCImageName(imagePrefix, version string, t model.ImageType) string {
+	if imagePrefix != "" && !strings.HasSuffix(imagePrefix,"/")  {
```
Comment (YourTechBud) [inline space-cli/cmd/utils/operations.go:100]:
```suggestion
	if !strings.HasSuffix(imagePrefix,"/")  {
```

File: space-cli/cmd/utils/operations.go [hunk 4]
```diff
@@ -93,6 +95,22 @@ func ChangeSelectedAccount(clusterName string) error {
...
+	}
+	switch t {
+	case model.ImageRunner:
+		return fmt.Sprintf("%s%s:%s", imagePrefix, "spaceuptech/runner", version)
+	case model.ImageGateway:
+		return fmt.Sprintf("%s%s:%s", imagePrefix, "spaceuptech/gateway", version)
+	default:
+		LogInfo("Invalid image type provided for getting sc image name with image prefix")
+		return ""
+	}
```
Comment (YourTechBud) [inline space-cli/cmd/utils/operations.go:111]:
```suggestion
	return fmt.Sprintf("%s%s:%s", imagePrefix, t, version)
```

File: space-cli/cmd/modules/operations/setup.go [hunk 5]
```diff
@@ -187,7 +187,7 @@ func Setup(username, key, config, version, secret, clusterName string, dev bool,
 	}{
 		{
 			name:           "gateway",
-			containerImage: fmt.Sprintf("%s:%s", "spaceuptech/gateway", version),
+			containerImage: utils.GetSCImageName(imagePrefix, version, model.ImageGateway),
 			containerName:  utils.GetScContainers(clusterName, "gateway"),
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/setup.go:191]:
```suggestion
			containerName:  utils.GetScContainers(clusterName, model.ImageGateway),
```

File: space-cli/cmd/modules/operations/setup.go [hunk 6]
```diff
@@ -205,7 +205,7 @@ func Setup(username, key, config, version, secret, clusterName string, dev bool,
 		{
 			// runner
 			name:           "runner",
-			containerImage: fmt.Sprintf("%s:%s", "spaceuptech/runner", version),
+			containerImage: utils.GetSCImageName(imagePrefix, version, model.ImageRunner),
 			containerName:  utils.GetScContainers(clusterName, "runner"),
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/setup.go:209]:
```suggestion
			containerName:  utils.GetScContainers(clusterName, model.ImageRunner),
```

File: space-cli/cmd/modules/operations/upgrade.go [hunk 7]
```diff
@@ -107,7 +108,7 @@ func Upgrade(clusterName string, version string) error {
 		portMapping    nat.PortMap
 	}{
 		{
-			containerImage: fmt.Sprintf("%s:%s", "spaceuptech/gateway", latestVersion),
+			containerImage: utils.GetSCImageName(imagePrefix, latestVersion, model.ImageGateway),
 			containerName:  utils.GetScContainers(clusterName, "gateway"),
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:112]:
```suggestion
			containerName:  utils.GetScContainers(clusterName, model.ImageGateway),
```

File: space-cli/cmd/modules/operations/upgrade.go [hunk 8]
```diff
@@ -14,11 +14,12 @@ import (
...
+	"github.com/spaceuptech/space-cloud/space-cli/cmd/model"
 	"github.com/spaceuptech/space-cloud/space-cli/cmd/utils"
 )
 
 // Upgrade upgrades the environment which has been setup
-func Upgrade(clusterName string, version string) error {
+func Upgrade(clusterName string, version string, imagePrefix string) error {
 
 	// Getting current version
 	result := make(map[string]interface{})
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:25]: Dont call utils.GetLatestVersion if the version flag was provided

File: space-cli/cmd/modules/operations/upgrade.go [hunk 9]
```diff
@@ -119,7 +120,7 @@ func Upgrade(clusterName string, version string) error {
 
 		{
 			// runner
-			containerImage: fmt.Sprintf("%s:%s", "spaceuptech/runner", latestVersion),
+			containerImage: utils.GetSCImageName(imagePrefix, latestVersion, model.ImageRunner),
 			containerName:  utils.GetScContainers(clusterName, "runner"),
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:124]:
```suggestion
			containerName:  utils.GetScContainers(clusterName, model.ImageRunner),
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: space-cli/cmd/modules/operations/setup.go [hunk 10]
```diff
@@ -143,7 +143,8 @@ func Setup(username, key, config, version, secret, clusterName string, dev bool,
 		"HOME_SECRETS_PATH=" + utils.GetMountTempSecretsDir(clusterName),
 		"HOSTS_FILE_PATH=" + utils.GetMountHostsFilePath(clusterName),
 		"ROUTING_FILE_PATH=" + "/routing-config.json",
-		"CLUSTER_ID=" + clusterID,
+		"CLUSTER_ID=" + clusterID,     // required for sending metrics
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/setup.go:146]: I dont think this is required

File: space-cli/cmd/modules/operations/upgrade.go [hunk 11]
```diff
@@ -27,13 +28,16 @@ func Upgrade(clusterName string, version string) error {
...
+	var err error
 	if version != "" {
 		latestVersion = version
+	} else {
+		// Getting latest version
+		latestVersion, err = utils.GetLatestVersion(currentVersion)
+		if err != nil {
+			return err
+		}
 	}
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:41]:
```suggestion
	latestVersion := version
	if latestVersion == "" {
		// Getting latest version
		latestVersion, err = utils.GetLatestVersion(currentVersion)
		if err != nil {
			return err
		}
	}
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: space-cli/cmd/modules/operations/setup.go (hunk 10 above)
Comment (sharadregoti) [reply space-cli/cmd/modules/operations/setup.go:146]: This flag is required, if you see the metric collection in runner. It uses this variable.

Gateway only gets an array of objects, aka values to be inserted. Each value contains the cluster id

File: runner/main.go [hunk 12]
```diff
@@ -62,6 +62,12 @@ func main() {
 					Usage:  "The cluster id to uniquely identify the cluster",
 					Value:  "first-cluster",
 				},
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply runner/main.go:64]: We can remove it but we have to set the environment variable.

[the remark this answers is not in the document]
Comment (YourTechBud) [reply runner/main.go:64]: So lets not remove it

File: space-cli/cmd/modules/operations/setup.go (hunk 10 above)
Comment (YourTechBud) [reply space-cli/cmd/modules/operations/setup.go:146]: Got it

File: space-cli/cmd/modules/operations/upgrade.go [hunk 13]
```diff
@@ -27,13 +28,13 @@ func Upgrade(clusterName string, version string) error {
...
 
-	// Getting latest version
-	latestVersion, err := utils.GetLatestVersion(currentVersion)
-	if err != nil {
-		return err
-	}
-	if version != "" {
-		latestVersion = version
+	latestVersion := version
+	if latestVersion == "" {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:32]:
```suggestion
	if latestVersion == "default" {
```

File: space-cli/cmd/modules/operations/upgrade.go [hunk 14]
```diff
@@ -27,13 +28,13 @@ func Upgrade(clusterName string, version string) error {
...
 
-	// Getting latest version
-	latestVersion, err := utils.GetLatestVersion(currentVersion)
-	if err != nil {
-		return err
-	}
-	if version != "" {
-		latestVersion = version
+	latestVersion := version
+	if latestVersion == "default" {
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/upgrade.go:32]:
```suggestion
	var err error
	latestVersion := version
	if latestVersion == "default" {
```

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**